File: back/api/guess.go
package api

import (
	"database/sql"
	"encoding/json"
	"net/http"
)

type GuessRequest struct {
	RoomID    string `json:"roomId"`
	Team      string `json:"team"`
	Members   int    `json:"members"`
	Guesses   []int  `json:"guesses"`
	ProductID int    `json:"productId"` // ← 単一商品のID
}

func RegisterGuessAPI(db *sql.DB) {
	http.HandleFunc("/api/guess-price", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type")
		if r.Method == "OPTIONS" {
			w.WriteHeader(http.StatusOK)
			return
		}

		var req GuessRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			http.Error(w, "invalid request", http.StatusBadRequest)
			return
		}

		// 重複チェック
		var exists int
		err := db.QueryRow(`SELECT COUNT(*) FROM guesses
			WHERE room_id = ? AND team = ? AND product_id = ?
		`, req.RoomID, req.Team, req.ProductID).Scan(&exists)
		if err != nil {
			http.Error(w, "database error", http.StatusInternalServerError)
			return
		}
		if exists > 0 {
			http.Error(w, "already submitted", http.StatusConflict) // 409 Conflict
			return
		}

		// 正解の価格をDBに問い合わせ
		var actualPrice int
		err = db.QueryRow("SELECT price FROM prices WHERE id = ?", req.ProductID).Scan(&actualPrice)
		if err != nil {
			http.Error(w, "price not found", http.StatusBadRequest)
			return
		}

		// 誤差の計算
		var totalError int
		for _, guess := range req.Guesses {
			diff := guess - actualPrice
			if diff < 0 {
				diff = -diff
			}
			totalError += diff
		}

		// 誤差の平均
		averageError := float64(totalError) / float64(len(req.Guesses))

		_, err = db.Exec("INSERT INTO guesses (room_id, team, product_id, error) VALUES (?, ?, ?, ?)", req.RoomID, req.Team, req.ProductID, averageError)
		if err != nil {
			http.Error(w, "failed to save guess", http.StatusInternalServerError)
			return
		}

		w.WriteHeader(http.StatusOK)
	})
}

File: back/api/guess.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from db import get_db
import re

router = APIRouter()

class GuessRequest(BaseModel):
    roomId: str
    team: str
    members: int
    guesses: list[int]
    productId: int

@router.post("/guess-price")
async def guess_price(req: GuessRequest):
    db = get_db()
    # duplicate check
    cur = db.execute(
        "SELECT COUNT(*) as cnt FROM guesses WHERE room_id=? AND team=? AND product_id=?",
        (req.roomId, req.team, req.productId)
    )
    if cur.fetchone()["cnt"] > 0:
        raise HTTPException(status_code=409, detail="already submitted")
    # Try to get the price from snack_prices table first
    cur = db.execute("SELECT price FROM snack_prices WHERE snack=?", (req.team,))
    row = cur.fetchone()
    
    if row:
        # Use the price from the snack_prices table
        actual = row["price"]
    else:
        # Fallback to the prices table
        cur = db.execute("SELECT price FROM prices WHERE id=?", (req.productId,))
        row = cur.fetchone()
        if not row:
            # If no price is found, use a default value
            # We can't call search_api directly here due to async/await complexity
            print(f"No price found for team: {req.team}")
            raise HTTPException(status_code=400, detail="price not found")
        else:
            actual = row["price"]
    # calculate error
    if not req.guesses:
        # Handle empty guesses list
        avg_error = 0
    else:
        total_error = sum(abs(g - actual) for g in req.guesses)
        avg_error = total_error / len(req.guesses)
    db.execute(
        "INSERT INTO guesses (room_id, team, product_id, error) VALUES (?,?,?,?)",
        (req.roomId, req.team, req.productId, avg_error)
    )
    db.commit()
    return {"status": "ok"}
